### PR TITLE
Add packet alert mode and monitoring of hardware originated drops

### PR DESCRIPTION
**Pull request by idosch**
Hi,

The drop monitor kernel module was recently extended with two new features which are expected to be merged to the 5.4 kernel. This pull request adds support for these new features in DropWatch.

The first feature was merged in [this](https://git.kernel.org/pub/scm/linux/kernel/git/davem/net-next.git/commit/?id=6e5ee483397aa02e5656191cb9f59e4095b4d249) kernel merge commit. It extends drop monitor with another alert mode in which dropped packets are notified to user space, instead of just a summary of recent drops. This enables users to perform a more detailed analysis of the dropped packets. Please refer to the kernel merge commit for example usage as well as to the individual commit messages in this pull request. The man page and help command were extended as well.

The second feature was merged in [this](https://git.kernel.org/pub/scm/linux/kernel/git/davem/net-next.git/commit/?id=83beee5a3aff0fb159b2fb4d0cac8f18a193417e) kernel merge commit. It allows drop monitor to monitor hardware originated drops, in addition to software originated drops.

Patches #1-#2 perform simple clean-ups.

Patch #3 synchronizes the kernel header file.

Patches #4-#8 add support for packet alert mode.

Patches #9-#12 allow DropWatch to monitor hardware originated drops.

I have tested this patched DropWatch version on a net-next kernel and a net kernel. Also tested unmodified DropWatch on a net-next kernel.

### CODE REVIEW

**Comment by idosch** [comment]
@dsahern David, I know you have some changes you wanted to make on top of this, so I'm copying you like I promised.

**Comment by dsahern** [comment]
Thanks for letting me know. My changes are a bit of hack to try out the new features. More thought is needed to better integrate packet analysis - eg., use libpcap to dump packets to a file. I doubt I will have time for something merge worthy in the next few months, so if someone else has the time and motivation please give it a try.

**Comment by idosch** [comment]
> Thanks for letting me know. My changes are a bit of hack to try out the new features. More thought is needed to better integrate packet analysis - eg., use libpcap to dump packets to a file. I doubt I will have time for something merge worthy in the next few months, so if someone else has the time and motivation please give it a try.

Hi David,

I thought about it and I'm not sure we need to add pcap support to DropWatch. Instead, I started writing a libpcap capture module that will allow tcpdump/wireshark to specifically capture drop monitor packets. It will basically open the netlink socket, configure drop monitor to switch to packet alert mode and start monitoring. The pcap filter will be run in user space on the NET_DM_ATTR_PAYLOAD data.

**Comment by nhorman** [comment]
comment requests?

File: src/main.c [hunk 1]
```diff
@@ -689,9 +706,19 @@ int enable_drop_monitor()
 
 	msg = alloc_netlink_msg(NET_DM_CMD_START, NLM_F_REQUEST|NLM_F_ACK, 0);
 
+	if (monitor_sw)
+		NLA_PUT_FLAG(msg->nlbuf, NET_DM_ATTR_SW_DROPS);
```
[the remark this answers is not in the document]
Comment (idosch) [reply src/main.c]: Neil, in order to be consistent with kernel code I will simply use the regular nla_put_* API and explicitly add the goto and label.

File: src/main.c [hunk 2]
```diff
@@ -566,6 +573,25 @@ int set_trunc_len()
...
 
+int set_queue_len()
+{
+	struct netlink_message *msg;
+
+	msg = alloc_netlink_msg(NET_DM_CMD_CONFIG, NLM_F_REQUEST|NLM_F_ACK, 0);
+	if (!msg)
+		return -ENOMEM;
+
+	NLA_PUT_U32(msg->nlbuf, NET_DM_ATTR_QUEUE_LEN, queue_len);
```
[the remark this answers is not in the document]
Comment (idosch) [reply src/main.c]: Will use nla_put_u32() instead

**Comment by idosch** [comment]
> comment requests?

Thanks for the review, Neil. I commented above. Hope it is fine.

File: src/main.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (nhorman) [reply src/main.c]: But this isn't kernel code, I'm asking for a reminder that the macro contains a goto, so I don't in the future wonder why theres an unused label there

[the remark this answers is not in the document]
Comment (idosch) [reply src/main.c]: > But this isn't kernel code, I'm asking for a reminder that the macro contains a goto, so I don't in the future wonder why theres an unused label there

Yes, I understand, but I'm saying that these macros were removed from the kernel in commit b3fe91c53a0a ("netlink: Delete all NLA_PUT*() macros."), so might make it easier for people flipping between kernel and user space code and avoid them here too.

[the remark this answers is not in the document]
Comment (nhorman) [reply src/main.c]: oh, sure, I'm sorry, yes, if you want to use the new macros that make the gotos explicit and clear, I'm totally fine with that

[the remark this answers is not in the document]
Comment (idosch) [reply src/main.c]: Great, thanks! I communicated the timestamps in a non y2038 safe way to user space and I have just sent a patch for that. I will re-submit the pull request with your comments addressed and the new timestamp format.

Speaking of re-submitting, should I open a new pull request or just push the changes to my branch?

[the remark this answers is not in the document]
Comment (nhorman) [reply src/main.c]: just push your changes to this pull request, no need for you to open a new one.  Thanks!

**Pull request merged**